cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRegisterer.java

```diff
@@ -156,7 +164,107 @@ public Pipeline registerPlugins(ETLConfig config, Class errorDatasetType, Datase
       throw new RuntimeException(e);
     }
 
-    return new Pipeline(sourcePluginId, sinksInfo, transformInfos);
+    Map<String, List<String>> connectionsMapString =
+      Maps.transformEntries(connectionsMap, new Maps.EntryTransformer<String, List<ETLStage>, List<String>>() {
+        @Override
+        public List<String> transformEntry(String key, List<ETLStage> stages) {
+          List<String> stageNames = new ArrayList<String>();
+          for (ETLStage stage : stages) {
+            stageNames.add(stage.getName());
+          }
+          return stageNames;
+        }
+      });
+
+    return new Pipeline(sourcePluginId, sinksInfo, transformInfos, connectionsMapString);
+  }
+
+  @VisibleForTesting
+  static Map<String, List<ETLStage>> validateConnections(ETLConfig config) {
+    Map<String, List<ETLStage>> mapSourceToNextStages = new HashMap<>();
```

Comment:
also, does the value need to be a `List<ETLStage>`? Doesn't a `List<String>` work just as well?